metric_metadata/cached/cached.go

```diff
@@ -111,37 +160,47 @@ func (c *CachedMetricMetadataAPI) CheckHealthy() error {
 	return c.metricMetadataAPI.CheckHealthy()
 }
 
-// getCachedTagSet is a thread-safe way to get the cached data for a metric (protected by a mutex)
-func (c *CachedMetricMetadataAPI) getCachedTagSet(metricKey api.MetricKey) CachedTagSetList {
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-	return c.getAllTagsCache[metricKey]
-}
-
-// setCachedTagSet is a thread-safe way to assign cached data (protected by a mutex).
-// if the given value is less recent than the stored value, nothing happens.
-func (c *CachedMetricMetadataAPI) replaceCachedTagSet(metricKey api.MetricKey, tagsets []api.TagSet, startTime time.Time) {
-	value := CachedTagSetList{
-		TagSets: tagsets,
-		Expiry:  startTime.Add(c.timeToLive),
-	}
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-	if c.getAllTagsCache[metricKey].Expiry.Before(value.Expiry) {
-		c.getAllTagsCache[metricKey] = value
+// fetchAndUpdateCachedTagSet updates the in-memory cache (asusming the update)
+// is newer than what is in the cache. The caller MUST have the lock on the item
+// in the cache already and will release that lock before returning
+func (c *CachedMetricMetadataAPI) fetchAndUpdateCachedTagSet(metricKey api.MetricKey, context metadata.Context) ([]api.TagSet, error) {
+	// No need to lock on getAllTagsCache since we already should hold the lock for
+	// the given item, so it should have already been written in a thread-safe way
+	item, ok := c.getAllTagsCache[metricKey]
+	if !ok {
+		return nil, errors.New("Missing item in cache")
 	}
-}
 
-// getAllTagsRaw performs a request to the underlying API and updates the cached values with whatever is returned.
-// Then it returns the result to the caller.
-// Note that getAllTagsRaw does not respect the request limit behavior- this must be enforced by the caller.
-func (c *CachedMetricMetadataAPI) getAllTagsRaw(metricKey api.MetricKey, context metadata.Context) ([]api.TagSet, error) {
-	startTime := time.Now()
+	item.wg.Add(1)
+	item.fetchError = nil
+	item.inflight = true
+	item.Unlock()
+
+	startTime := c.clock.Now()
 	tagsets, err := c.metricMetadataAPI.GetAllTags(metricKey, context)
 	if err != nil {
+		item.Lock()
+		item.fetchError = err
+		item.wg.Done()
+		item.inflight = false
+		item.Unlock()
```

Comment:
Yeah realized this is duplicated, so just moved it right after the `GetAllTags` call.